### DISCUSSION TITLE
Combining data and logic

### DISCUSSION THREAD

**Discussion by canmingir** [NucleoidAI/Nucleoid#28] [Thinkers Club]
Engineering is all about eliminating/reducing moving pieces with better designing things. Nikola Tesla's valve is perfect example for that. Due to imperative programming, data and logic have to be separated out, and there is no way out.

In search of finding better engineering design in software architecture, in fact we already have very stable libraries can take of problems like connection/thread pools, locking systems etc. but I think we can do better than that.

It is impossible to do with imperative programming, but declarative programming as a runtime system may solve and combine data and logic under the same process. Immediate advantages that switching up the programming model gives us less code, less library, less complexity means less problem (I am pretty sure different problems tho). In my opinion, declarative runtimes are electric cars of software development, has much more potentials that waiting to be discovered.

**Comment by francisco-giancarelli-crombie** [comment]
Yes, I agree. But I think, if you want programmers to use Nucleoid I think is not just a matter of being able to automatically manage data, it must be at least close to the performance and scalability of what a custom data layer would have. If you are able to achieve that, then Nucleoid will be a powerful option.
